pkg/util/retry/retry.go

```diff
@@ -0,0 +1,126 @@
+/*
+Copyright 2022 Codenotary Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package retry
+
+import (
+	"context"
+	"errors"
+	"time"
+)
+
+var (
+	// DefaultRetry is backoff with jitter retrier strategy.
+	DefaultRetry = Backoff{
+		Steps:   5,
+		retrier: NewExponentialRetrier(2*time.Millisecond, 10*time.Millisecond, 2.0, 0),
+	}
+
+	// DefaultConstant is backoff with constant retrier strategy.
+	DefaultConstant = Backoff{
+		Steps:   5,
+		retrier: NewConstantRetrier(10 * time.Millisecond),
+	}
+)
+
+var (
+	// ErrRetryAgain is an error used to retry on a fail condition
+	ErrRetryAgain = errors.New("retry again")
+	// ErrNoRetry is an error used to not retry on error
+	ErrNoRetry = errors.New("do not retry on error")
+)
+
+// Retriable returns true if the error has to be retried
+type Retriable func(error) bool
```

Comment:
I would go for a single method for determination if something is retryable or not.

My suggestion here is:

* if the returned value is `ErrRetryAgain` - this should be treated as now - just retry the step
* introduce something like `errNoRetry` struct (and a method `ErrNoRetry` to create its instance) that wraps another error - that way in retryable step one can return an error that will break the chain and will not be retried
* optional: introduce something like `errStop` struct that will stop the chain but the result of chain execution will not be an error - this one is not currently needed thus I'm not sure if it's worth implementing it now (and it will make the chain implementation a bit more complex) - but the main idea is that it should be relatively easy to add if needed